interfaces/cython/cantera/drawnetwork.py

```diff
@@ -0,0 +1,474 @@
+                     node_attr=None, edge_attr=None, heat_flow_attr=None,
+                     mass_flow_attr=None):
+    """
+    Draw `ReactorNet` object as ``graphviz.graphs.DiGraph``. Connecting flow
+    controllers and walls are depicted as arrows.
+
+    :param n:
+        `ReactorNet` object
+    :param print_state:
+        Whether state information of the reactors is printed into each node.
+        Defaults to ``False``.
+    :param species:
+        If ``print_state`` is ``True``, define how species are to be printed.
+        Options are ``'X'`` and ``'Y'`` for mole and mass fractions of all
+        species, respectively, or an iterable that contains the desired species
+        names as strings. Defaults to ``None``.
+    :param graph_attr:
+        Attributes to be passed to the ``graphviz.Digraph`` function that
+        control the general appearance of the drawn network.
+        See https://graphviz.org/docs/graph/ for a list of all usable
+        attributes.
+    :param node_attr:
+        Attributes to be passed to the ``node`` method invoked to draw the
+        reactor. ``node_attr`` defined in the reactor object itself have
+        priority.
+        See https://graphviz.org/docs/nodes/ for a list of all usable
+        attributes.
+    :param edge_attr:
+        Attributes to be passed to the ``edge`` method invoked to draw
+        reactor connections. ``edge_attr`` defined in the connection objects
+        (subclasses of `FlowDevice` or walls) themselve have priority.
+        See https://graphviz.org/docs/edges/ for a list of all usable
+        attributes.
+    :param heat_flow_attr:
+        Same as `edge_attr` but only applied to edges representing walls.
+    :param mass_flow_attr:
+        Same as `edge_attr` but only applied to edges representing
+        `FlowDevice` objects.
+    :return:
+        ``graphviz.graphs.BaseGraph`` object with reactor net.
+
+    """
+    kwargs = locals()
+    kwargs.pop('n',)
+    dot = _graphviz.Digraph(graph_attr=graph_attr)
+
+    # collect elements as set to avoid duplicates
+    reactors = set(n.reactors)
+    connections = set()
+    drawn_reactors = set()
+
+    reactor_groups = {}
+    for r in reactors:
+        if r.groupname not in reactor_groups:
+            reactor_groups[r.groupname] = set()
+        reactor_groups[r.groupname].add(r)
+
+    reactor_groups.pop("", None)
+    if reactor_groups:
+        for name, group in reactor_groups.items():
+            sub = _graphviz.Digraph(name=f"cluster_{name}",
+                                    graph_attr=graph_attr)
+            for r in group:
+                _draw_reactor(r, sub, **kwargs)
+                drawn_reactors.add(r)
+                connections.update(r.walls + r.inlets + r.outlets)
+                for surface in r.surfaces:
+                    _draw_surface(surface, sub, **kwargs)
+            sub.attr(label=name)
+            dot.subgraph(sub)
+    reactors.difference_update(drawn_reactors)
```

Comment:
I'd recommend the operator notation here (and a couple other places):
```suggestion
    reactors -= drawn_reactors
```